### PR TITLE
Fix concurrency bug in test suite

### LINKED ISSUE #6832: `DuplicateExecutable` test randomly breaks other `Fail/` tests
On latest `master` (283730b), running `make test` fails non-deterministically.

The console pauses several times and displays diffs such as
```diff
1.actual
index e83670021..a3a9e853d 100644
--- a/tmp/nix-shell.qnWrHw/AbstractModuleMacro524397-0.golden
+++ b/tmp/nix-shell.qnWrHw/AbstractModuleMacro524397-1.actual
@@ -1,5 +1,3 @@
-AbstractModuleMacro.agda:7,1-8,19
-warning: -W[no]UselessAbstract
-Using abstract here has no effect. Abstract applies to only
-definitions like data definitions, record type definitions and
-function clauses.
+Duplicate entries for executable 'echo' in DuplicateExecutable/executables:
+  2: /bin/echo
+  3: /bin/echo
```
or
```diff
diff --git a/tmp/nix-shell.qnWrHw/BadForcedIrrelevance524397-2.golden b/tmp/nix-shell.qnWrHw/BadForcedIrrelevance524397-3.actual
index dc1787cd6..a3a9e853d 100644
--- a/tmp/nix-shell.qnWrHw/BadForcedIrrelevance524397-2.golden
+++ b/tmp/nix-shell.qnWrHw/BadForcedIrrelevance524397-3.actual
@@ -1,3 +1,3 @@
-BadForcedIrrelevance.agda:11,16-17
-Variable n is declared erased, so it cannot be used here
-when checking that the expression n has type Nat
+Duplicate entries for executable 'echo' in DuplicateExecutable/executables:
+  2: /bin/echo
+  3: /bin/echo
(END)
```
the incoming fragment is always for the `DuplicateExecutable` test. The other test varies, but is always from the `Fail/` suite.

When running `make test -j1`, exactly 1 other test is affected, but without the `-j1` flag the number of affected tests varies.

I am using GNU Make 4.4.1, cabal 3.8.1.0, and ghc 9.2.7, all provided by `nix develop` in this repo's root.
OS is Ubuntu 22.04.3 LTS x86_64

### PR DESCRIPTION
**Pull request by lawcho**
Fixes #6832

### CODE REVIEW

File: src/full/Agda/Utils/Environment.hs [hunk 1]
```diff
@@ -25,6 +28,13 @@ expandVars home env s = concatMap repl $ tokens s
     repl (Var x) = fromMaybe "" $ lookup x env
     repl (Str s) = s
 
+-- | Expand a telescope of environment variables
+--   (each value may refer to variables earlier in the list)
+expandEnvVarTelescope :: String -> [(String, String)] -> [(String, String)]
+expandEnvVarTelescope home vs0 = reverse $ foldr  -- compensate for reverse below
+  (\(var,val) acc -> (var, expandVars home acc val):acc)
+  [] (reverse vs0)  -- reverse because earlier vals must be processed first
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]: Isn't a `foldr . reverse` a `foldl`?

(Update: As opposed to a `reverse . foldr` which you corrected.)

File: test/Utils.hs [hunk 2]
```diff
@@ -66,13 +66,14 @@ printProgramResult = printProcResult . fromProgramResult
 type AgdaArgs = [String]
 
 
-readAgdaProcessWithExitCode :: AgdaArgs -> Text
+readAgdaProcessWithExitCode :: Maybe [(String, String)]
+                            -> AgdaArgs -> Text
                             -> IO (ExitCode, Text, Text)
-readAgdaProcessWithExitCode args inp = do
+readAgdaProcessWithExitCode env args inp = do
```
Comment (andreasabel) [inline test/Utils.hs]: Should the following `getEnvAgdaArgs` take `env` into account?
Either way, please leave a comment _in the code_ clarifying this.

File: src/full/Agda/Utils/Environment.hs [hunk 3]
```diff
@@ -25,6 +28,13 @@ expandVars home env s = concatMap repl $ tokens s
     repl (Var x) = fromMaybe "" $ lookup x env
     repl (Str s) = s
 
+-- | Expand a telescope of environment variables
+--   (each value may refer to variables earlier in the list)
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]: Please document (somewhere suitable) that the `.flags` file can be a such a telescope (i.e. have such dependencies).

**Review by andreasabel — COMMENTED** [review]
Thanks!
Not manipulating a global env is indeed much better than what we have now!

I have a few comments.

File: src/full/Agda/Utils/Environment.hs (hunk 1 above)
Comment (lawcho) [reply src/full/Agda/Utils/Environment.hs]: I see two constraints:

* Expansion must happen left-to-right
* The output list must have the same order as the input list

This works: `reverse . foldr f [] . reverse`
This also works: `reverse . foldl (flip f) []`

File: test/Utils.hs (hunk 2 above)
Comment (lawcho) [reply test/Utils.hs]: Made `getEnvAgdaArgs` & related helpers pure

File: src/full/Agda/Utils/Environment.hs (hunk 3 above)
Comment (lawcho) [reply src/full/Agda/Utils/Environment.hs]: Documented in `HACKING.md`

File: src/full/Agda/Utils/Environment.hs [hunk 4]
```diff
@@ -28,12 +29,14 @@ expandVars home env s = concatMap repl $ tokens s
...
+type EnvVars = [(String, String)]
+
 -- | Expand a telescope of environment variables
 --   (each value may refer to variables earlier in the list)
-expandEnvVarTelescope :: String -> [(String, String)] -> [(String, String)]
-expandEnvVarTelescope home vs0 = reverse $ foldr  -- compensate for reverse below
-  (\(var,val) acc -> (var, expandVars home acc val):acc)
-  [] (reverse vs0)  -- reverse because earlier vals must be processed first
+expandEnvVarTelescope :: String -> EnvVars -> EnvVars
+expandEnvVarTelescope home = reverse . foldl  -- compensate for reverse below
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]: Maybe the comment would be now: "compensate for reversal done by foldl".

**Review by andreasabel — APPROVED** [review]

File: src/full/Agda/Utils/Environment.hs (hunk 4 above)
Comment (lawcho) [reply src/full/Agda/Utils/Environment.hs]: Updated comment

File: src/full/Agda/Utils/Environment.hs [hunk 5]
```diff
@@ -25,6 +29,15 @@ expandVars home env s = concatMap repl $ tokens s
     repl (Var x) = fromMaybe "" $ lookup x env
     repl (Str s) = s
 
+-- | List of environment variable bindings
+type EnvVars = [(String, String)]
+
+-- | Expand a telescope of environment variables
+--   (each value may refer to variables earlier in the list)
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]:
```suggestion
--   (each value may refer to variables earlier in the list).
```

File: src/full/Agda/Utils/Environment.hs [hunk 6]
```diff
@@ -25,6 +29,15 @@ expandVars home env s = concatMap repl $ tokens s
     repl (Var x) = fromMaybe "" $ lookup x env
     repl (Str s) = s
 
+-- | List of environment variable bindings
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]:
```suggestion
-- | List of environment variable bindings.
```

File: src/full/Agda/Utils/Environment.hs [hunk 7]
```diff
@@ -25,6 +29,15 @@ expandVars home env s = concatMap repl $ tokens s
...
+type EnvVars = [(String, String)]
+
+-- | Expand a telescope of environment variables
+--   (each value may refer to variables earlier in the list)
+expandEnvVarTelescope :: String -> EnvVars -> EnvVars
+expandEnvVarTelescope home = reverse . foldl  -- foldl reverses, so re-reverse afterwards
+  (\acc (var,val) -> (var, expandVars home acc val):acc) []
```
Comment (andreasabel) [inline src/full/Agda/Utils/Environment.hs]:
```suggestion
  (\ acc (var, val) -> (var, expandVars home acc val) : acc) []
```

File: test/Utils.hs [hunk 8]
```diff
@@ -66,13 +66,16 @@ printProgramResult = printProcResult . fromProgramResult
 type AgdaArgs = [String]
 
 
-readAgdaProcessWithExitCode :: AgdaArgs -> Text
+readAgdaProcessWithExitCode :: Maybe EnvVars  --- extra env vars, unexpanded
```
Comment (andreasabel) [inline test/Utils.hs]:
```suggestion
readAgdaProcessWithExitCode :: Maybe EnvVars  -- ^ Extra environment variables, unexpanded.
```

**Review by andreasabel — COMMENTED** [review]
I suggest some cosmetic fixes.
- Pushed a commit with cosmetics in the HACKING file---but actually wanted to use the suggestion feature.
- Used the suggestion feature.  You can batch the suggestions and then apply them together.

File: test/Utils.hs [hunk 9]
```diff
@@ -66,13 +66,16 @@ printProgramResult = printProcResult . fromProgramResult
...
-  agdaBin <- getAgdaBin
-  envArgs <- getEnvAgdaArgs
+readAgdaProcessWithExitCode extraEnv args inp = do
+  origEnv <- getEnvironment
+  home <- getHomeDirectory
+  let env = expandEnvVarTelescope home $ maybe origEnv (origEnv ++) extraEnv
+  let envArgs = maybe [] words $ lookup "AGDA_ARGS" env
   -- hPutStrLn stderr $ unwords $ agdaBin : envArgs ++ args
-  let agdaProc = (proc agdaBin (envArgs ++ args)) { create_group = True }
+  let agdaProc = (proc (getAgdaBin env) (envArgs ++ args)) { create_group = True , env = Just env }
```
Comment (andreasabel) [inline test/Utils.hs:78]: @lawcho: Could this change somehow have triggered 
- #6905?

Comment (andreasabel) [reply test/Utils.hs:78]: So this is likely a bug in the `process` library that only surfaces on `macOS`.  The fix of this bug suggests that setting `cwd` here would workaround this bug, see https://github.com/haskell/process/pull/296/files.

Comment (andreasabel) [reply test/Utils.hs:78]: My experiments confirm that indeed this PR surfaced the bug in `process` currently affecting GHCs 9.2.8, 9.4.7, and 9.6.3.  Sorry, you stepped on a mine here, you take no blame.

**Pull request merged**